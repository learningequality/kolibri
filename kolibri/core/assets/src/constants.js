import invert from 'lodash/invert';
import Subjects from 'kolibri-constants/labels/Subjects';
// coach-facing
export { default as ContentNodeResourceType } from 'kolibri-constants/labels/ResourceType';
export { default as LearningActivities } from 'kolibri-constants/labels/LearningActivities';
export { default as AccessibilityCategories } from 'kolibri-constants/labels/AccessibilityCategories';
// Used to categorize the level or audience of content
export { default as ContentLevels } from 'kolibri-constants/labels/Levels';
export { default as ResourcesNeededTypes } from 'kolibri-constants/labels/Needs';
export { default as Categories } from 'kolibri-constants/labels/Subjects';

export const UserKinds = {
  ADMIN: 'admin',
  COACH: 'coach',
  LEARNER: 'learner',
  SUPERUSER: 'superuser',
  ANONYMOUS: 'anonymous',
  ASSIGNABLE_COACH: 'classroom assignable coach',
  CAN_MANAGE_CONTENT: 'can manage content',
};

export const CollectionKinds = {
  CLASSROOM: 'classroom',
  LEARNERGROUP: 'learnergroup',
  ADHOCLEARNERSGROUP: 'adhoclearnersgroup',
};

export const ContentNodeKinds = {
  AUDIO: 'audio',
  DOCUMENT: 'document',
  VIDEO: 'video',
  EXERCISE: 'exercise',
  TOPIC: 'topic',
  HTML5: 'html5',
  CHANNEL: 'channel', // e.g. a root topic
  EXAM: 'exam',
  LESSON: 'lesson',
  CLASSROOM: 'CLASSROOM',
  ACTIVITY: 'ACTIVITY',
  SLIDESHOW: 'slideshow',
  BOOKMARK: 'bookmark',
};

export const CategoriesLookup = invert(Subjects);

export const AllCategories = 'all_categories';

export const NoCategories = 'no_categories';

// used internally on the client as a hack to allow content-icons to display users
export const USER = 'user';

export const InteractionTypes = {
  hint: 'hint',
  answer: 'answer',
  error: 'error',
};

// enum values for `assessmentdata.mastery_model.type` field
// from le-utils/le_utils/constants/exercises.py
export const MasteryModelTypes = Object.freeze({
  do_all: 'do_all',
  num_correct_in_a_row_2: 'num_correct_in_a_row_2',
  num_correct_in_a_row_3: 'num_correct_in_a_row_3',
  num_correct_in_a_row_5: 'num_correct_in_a_row_5',
  num_correct_in_a_row_10: 'num_correct_in_a_row_10',
  m_of_n: 'm_of_n',
});

export const MasteryModelGenerators = {
  [MasteryModelTypes.do_all]: assessmentIds => ({
    m: assessmentIds.length,
    n: assessmentIds.length,
  }),
  [MasteryModelTypes.num_correct_in_a_row_2]: () => ({ m: 2, n: 2 }),
  [MasteryModelTypes.num_correct_in_a_row_3]: () => ({ m: 3, n: 3 }),
  [MasteryModelTypes.num_correct_in_a_row_5]: () => ({ m: 5, n: 5 }),
  [MasteryModelTypes.num_correct_in_a_row_10]: () => ({ m: 10, n: 10 }),
  [MasteryModelTypes.m_of_n]: (assessmentIds, masteryModel) => masteryModel,
};

// How many points is a completed content item worth?
export const MaxPointsPerContent = 500;

export const LoginErrors = {
  PASSWORD_MISSING: 'PASSWORD_MISSING',
  INVALID_CREDENTIALS: 'INVALID_CREDENTIALS',
  PASSWORD_NOT_SPECIFIED: 'PASSWORD_NOT_SPECIFIED',
  USER_NOT_FOUND: 'USER_NOT_FOUND',
};

export const PermissionTypes = {
  SUPERUSER: 'SUPERUSER',
  LIMITED_PERMISSIONS: 'LIMITED_PERMISSIONS',
  NO_DEVICE_PERMISSIONS: 'NO_DEVICE_PERMISSIONS',
};

export const SIGNED_OUT_DUE_TO_INACTIVITY = 'SIGNED_OUT_DUE_TO_INACTIVITY';

export const UPDATE_MODAL_DISMISSED = 'UPDATE_MODAL_DISMISSED';

export const NavComponentSections = {
  ACCOUNT: 'account',
};

export const SyncStatus = {
  RECENTLY_SYNCED: 'RECENTLY_SYNCED',
  SYNCING: 'SYNCING',
  QUEUED: 'QUEUED',
  UNABLE_TO_SYNC: 'UNABLE_TO_SYNC',
  NOT_RECENTLY_SYNCED: 'NOT_RECENTLY_SYNCED',
  UNABLE_OR_NOT_SYNCED: 'UNABLE_OR_NOT_SYNCED',
  NOT_CONNECTED: 'NOT_CONNECTED',
};

export const ERROR_CONSTANTS = {
  // These are an exact copy of the python module kolibri.core.error_constants
  // and should be kept in sync.
  // 400 error based constants
  USERNAME_ALREADY_EXISTS: 'USERNAME_ALREADY_EXISTS',
  USER_ALREADY_IN_GROUP_IN_CLASS: 'USER_ALREADY_IN_GROUP_IN_CLASS',
  MISSING_PASSWORD: 'MISSING_PASSWORD',
  MAX_LENGTH: 'MAX_LENGTH',
  INVALID: 'INVALID',
  UNIQUE: 'UNIQUE',
  INVALID_NETWORK_LOCATION_FORMAT: 'INVALID_NETWORK_LOCATION_FORMAT',
  NETWORK_LOCATION_NOT_FOUND: 'NETWORK_LOCATION_NOT_FOUND',
  ALREADY_REGISTERED_FOR_COMMUNITY: 'ALREADY_REGISTERED_FOR_COMMUNITY',
  // 401 error constants
  INVALID_CREDENTIALS: 'INVALID_CREDENTIALS',
  // 404 error constants
  NOT_FOUND: 'NOT_FOUND',
  INVALID_KDP_REGISTRATION_TOKEN: 'INVALID_KDP_REGISTRATION_TOKEN',
  // 403 error constants
  PERMISSION_DENIED: 'PERMISSION_DENIED',
  NOT_AUTHENTICATED: 'NOT_AUTHENTICATED',
  PASSWORD_NOT_SPECIFIED: 'PASSWORD_NOT_SPECIFIED',
  AUTHENTICATION_FAILED: 'AUTHENTICATION_FAILED',
  DEVICE_LIMITATIONS: 'DEVICE_LIMITATIONS',
};

export const DemographicConstants = {
  NOT_SPECIFIED: 'NOT_SPECIFIED',
  DEFERRED: 'DEFERRED',
};

// See FacilityUser model
export const FacilityUserGender = {
  MALE: 'MALE',
  FEMALE: 'FEMALE',
  ...DemographicConstants,
};
