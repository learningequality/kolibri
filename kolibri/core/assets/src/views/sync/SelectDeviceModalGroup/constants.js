/**
 * @see kolibri/core/discovery/models.py:ConnectionStatus
 */
export const ConnectionStatus = {
  Unknown: 'Unknown',
  ConnectionFailure: 'ConnectionFailure',
  ResponseTimeout: 'ResponseTimeout',
  ResponseFailure: 'ResponseFailure',
  InvalidResponse: 'InvalidResponse',
  Conflict: 'Conflict',
};

export const UnreachableConnectionStatuses = [
  ConnectionStatus.ConnectionFailure,
  ConnectionStatus.ResponseTimeout,
];
