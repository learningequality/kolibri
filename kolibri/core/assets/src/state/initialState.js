import { UserKinds } from '../constants';

const baseLoggingState = {
  summary: { progress: 0 },
  session: {},
  mastery: {},
  attempt: {},
};

export const baseSessionState = {
  id: undefined,
  username: '',
  full_name: '',
  user_id: undefined,
  facility_id: undefined,
  kind: [UserKinds.ANONYMOUS],
  can_manage_content: false,
};

const baseConnectionState = {
  connected: true,
  reconnectTime: null,
};

// core state is namespaced, and merged with a particular app's state
export const initialState = {
  core: {
    error: '',
    loading: true,
    title: '',
    pageSessionId: 0,
    session: baseSessionState,
    loginError: null,
    signInBusy: false,
    logging: baseLoggingState,
    totalProgress: null,
    channels: {
      list: [],
      currentId: null,
    },
    facilityConfig: {},
    facilities: [],
    connection: baseConnectionState,
    snackbarIsVisible: false,
    snackbarOptions: {},
  },
};
