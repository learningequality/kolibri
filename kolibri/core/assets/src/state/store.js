import Vuex from 'vuex';
import Vue from 'vue';
import { UserKinds } from '../constants';

Vue.use(Vuex);

const baseLoggingState = {
  summary: { progress: 0 },
  session: {},
  mastery: {},
  attempt: {},
};

const baseSessionState = {
  id: undefined,
  username: '',
  full_name: '',
  user_id: undefined,
  facility_id: undefined,
  kind: [UserKinds.ANONYMOUS],
  can_manage_content: false,
};

const baseConnectionState = {
  connected: true,
  reconnectTime: null,
};

// core state is namespaced, and merged with a particular app's state
export const initialState = {
  core: {
    error: '',
    loading: true,
    title: '',
    pageSessionId: 0,
    session: baseSessionState,
    loginError: null,
    signInBusy: false,
    logging: baseLoggingState,
    totalProgress: null,
    channels: {
      list: [],
      currentId: null,
    },
    facilityConfig: {},
    facilities: [],
    connection: baseConnectionState,
    snackbarIsVisible: false,
    snackbarOptions: {},
  },
};

export const coreMutations = {
  CORE_SET_SESSION(state, value) {
    Object.assign(state.core.session, value);
  },
  CORE_SET_FACILITY_CONFIG(state, facilityConfig) {
    state.core.facilityConfig = facilityConfig;
  },
  CORE_SET_FACILITIES(state, facilities) {
    state.core.facilities = facilities;
  },
  // Makes settings for wrong credentials 401 error
  CORE_SET_LOGIN_ERROR(state, value) {
    state.core.loginError = value;
  },
  CORE_SET_SIGN_IN_BUSY(state, isBusy) {
    state.core.signInBusy = isBusy;
  },
  CORE_CLEAR_SESSION(state) {
    Object.assign(state.core.session, baseSessionState);
  },
  CORE_SET_PAGE_LOADING(state, value) {
    const update = { loading: value };
    if (value) {
      Object.assign(update, { pageSessionId: state.core.pageSessionId + 1 });
    }
    Object.assign(state.core, update);
  },
  CORE_SET_ERROR(state, error) {
    state.core.error = error;
  },
  CORE_SET_TITLE(state, title) {
    state.core.title = title;
  },
  CORE_SET_CONNECTED(state, connected) {
    state.core.connection.connected = connected;
  },
  CORE_SET_RECONNECT_TIME(state, reconnectTime) {
    state.core.connection.reconnectTime = reconnectTime;
  },
  SET_LOGGING_SUMMARY_STATE(state, summaryState) {
    state.core.logging.summary = summaryState;
  },
  SET_LOGGING_SUMMARY_ID(state, summaryId) {
    state.core.logging.summary.id = summaryId;
  },
  SET_LOGGING_SESSION_ID(state, sessionId) {
    state.core.logging.session.id = sessionId;
  },
  SET_LOGGING_SESSION_STATE(state, sessionState) {
    state.core.logging.session = sessionState;
  },
  SET_LOGGING_PROGRESS(state, sessionProgress, summaryProgress) {
    state.core.logging.session.progress = sessionProgress;
    state.core.logging.summary.progress = summaryProgress;
  },
  SET_LOGGING_COMPLETION_TIME(state, time) {
    state.core.logging.summary.completion_timestamp = time;
  },
  SET_LOGGING_TIME(state, sessionTime, summaryTime, currentTime) {
    state.core.logging.session.end_timestamp = currentTime;
    state.core.logging.summary.end_timestamp = currentTime;
    state.core.logging.session.time_spent = sessionTime;
    state.core.logging.summary.time_spent = summaryTime;
  },
  SET_LOGGING_PENDING(state, summaryPending, sessionPending) {
    state.core.logging.summary.pending_create = summaryPending;
    state.core.logging.session.pending_create = sessionPending;
  },
  SET_LOGGING_THRESHOLD_CHECKS(state, progress, timeSpent) {
    state.core.logging.session.total_time_at_last_save = timeSpent;
    state.core.logging.session.progress_at_last_save = progress;
  },
  SET_LOGGING_MASTERY_STATE(state, masteryState) {
    state.core.logging.mastery = masteryState;
  },
  SET_LOGGING_MASTERY_COMPLETE(state, completetime) {
    state.core.logging.mastery.complete = true;
    state.core.logging.mastery.completion_timestamp = completetime;
  },
  SET_LOGGING_ATTEMPT_STATE(state, attemptState) {
    state.core.logging.attempt = attemptState;
  },
  SET_LOGGING_ATTEMPT_STARTTIME(state, starttime) {
    state.core.logging.attempt.start_timestamp = starttime;
  },
  UPDATE_LOGGING_ATTEMPT_INTERACTION_HISTORY(state, action) {
    state.core.logging.attempt.interaction_history.push(action);
  },
  UPDATE_LOGGING_MASTERY(state, currentTime, correct, firstAttempt, hinted, error) {
    if (firstAttempt) {
      state.core.logging.mastery.totalattempts += 1;
      state.core.logging.mastery.pastattempts.unshift({ correct, hinted, error });
    }
    state.core.logging.mastery.end_timestamp = currentTime;
  },
  UPDATE_LOGGING_ATTEMPT(
    state,
    { currentTime, correct, firstAttempt, complete, hinted, answerState, simpleAnswer, error }
  ) {
    if (complete) {
      state.core.logging.attempt.completion_timestamp = currentTime;
      state.core.logging.attempt.complete = true;
    } else {
      state.core.logging.attempt.completion_timestamp = null;
      state.core.logging.attempt.complete = false;
    }
    state.core.logging.attempt.end_timestamp = currentTime;
    let starttime = state.core.logging.attempt.start_timestamp;
    if (typeof starttime === 'string') {
      starttime = new Date(starttime);
    }
    state.core.logging.attempt.time_spent = currentTime - starttime;
    if (firstAttempt) {
      // Can only get it correct on the first try.
      state.core.logging.attempt.correct = correct;
      state.core.logging.attempt.hinted = hinted;
      state.core.logging.attempt.answer = answerState;
      state.core.logging.attempt.simple_answer = simpleAnswer;
      state.core.logging.attempt.error = error;
    } else if (state.core.logging.attempt.correct < 1) {
      // Only set hinted if attempt has not already been marked as correct
      // and set it to true if now true, but leave as true if false.
      state.core.logging.attempt.hinted = state.core.logging.attempt.hinted || hinted;
    }
  },
  SET_EMPTY_LOGGING_STATE(state) {
    state.core.logging.summary = { progress: 0 };
    state.core.logging.session = {};
    state.core.logging.mastery = {};
    state.core.logging.attempt = {};
  },
  SET_TOTAL_PROGRESS(state, progress) {
    state.core.totalProgress = progress;
  },
  SET_CORE_CHANNEL_LIST(state, channelList) {
    state.core.channels.list = channelList;
  },

  CORE_CREATE_SNACKBAR(state, snackbarOptions = {}) {
    // reset
    state.core.snackbarIsVisible = false;
    state.core.snackbarOptions = {};
    // set new options
    state.core.snackbarIsVisible = true;
    state.core.snackbarOptions = snackbarOptions;
  },

  CORE_CLEAR_SNACKBAR(state) {
    state.core.snackbarIsVisible = false;
    state.core.snackbarOptions = {};
  },
  CORE_SET_SNACKBAR_TEXT(state, text) {
    state.core.snackbarOptions.text = text;
  },
  RESET_EXAM_ATTEMPT_LOGS(state) {
    state.examAttemptLogs = {};
  },
};

const store = new Vuex.Store({});

export default store;

store.registerModule = ({ state, mutations } = { state: {}, mutations: {} }) => {
  if (store.__initialized) {
    throw new Error(
      'The store has already been initialized, dynamic initalization is not currently available'
    );
  }
  store.hotUpdate({ mutations: Object.assign(mutations, coreMutations) });
  store.replaceState(Object.assign(state, initialState));
  store.__initialized = true;
};

store.factory = ({ state, mutations } = { state: {}, mutations: {} }) => {
  store.__initialized = false;
  store.registerModule({ state, mutations });
  return store;
};
