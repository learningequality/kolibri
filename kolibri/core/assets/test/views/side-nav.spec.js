import times from 'lodash/times';
import { mount } from '@vue/test-utils';
import navComponents from 'kolibri.utils.navComponents';
import { UserKinds, NavComponentSections } from 'kolibri.coreVue.vuex.constants';
import SideNav from '../../src/views/SideNav';
import logoutSideNavEntry from '../../src/views/LogoutSideNavEntry';
import LearnOnlyDeviceNotice from '../../src/views/LearnOnlyDeviceNotice';
import SyncStatusDisplay from '../../src/views/SyncStatusDisplay';
import { stubWindowLocation } from 'testUtils'; // eslint-disable-line

import { coreStoreFactory as makeStore } from '../../src/state/store';

jest.mock('kolibri.urls');

function createWrapper({ navShown = true, headerHeight = 20, width = 100 } = {}, data = {}) {
  return mount(SideNav, {
    propsData: {
      navShown,
      headerHeight,
      width,
    },
    data() {
      return { ...data };
    },
    store: makeStore(),
  });
}

function setUserKind(store, userKind) {
  let canManageContent = false;
  if (userKind == UserKinds.CAN_MANAGE_CONTENT) {
    userKind = UserKinds.LEARNER;
    canManageContent = true;
  }
  store.commit('CORE_SET_SESSION', {
    id: 'test',
    username: 'test',
    full_name: 'testing test',
    user_id: 'test_id',
    facility_id: 'a real school',
    kind: [userKind],
    can_manage_content: canManageContent,
  });
}

const url = '/test/url';
const label = 'label1';
const label2 = 'label2';
const icon = 'library';
const role = UserKinds.LEARNER;

function createAndRegisterComponent(name, url, label, icon, role, priority, section) {
  const config = {
    name: name,
    url: url,
    label: label,
    icon: icon,
    role: role,
    priority: priority,
    section: section,
    bottomBar: true,
  };
  navComponents.register(config);
  return config;
}

function emptyNavComponents(n = 1) {
  times(n, () => {
    navComponents.pop();
  });
  expect(navComponents).toHaveLength(0);
}

describe('side nav component', () => {
  stubWindowLocation(beforeAll, afterAll);

  it('should be hidden if navShown is false', () => {
    const wrapper = createWrapper({ navShown: false });
    expect(wrapper.find('.side-nav').element).not.toBeVisible();
  });
  it('should show nothing if no components are added and user is not logged in', () => {
    expect(navComponents).toHaveLength(0);
    const wrapper = createWrapper();
    expect(wrapper.find('a.ui-menu-option:not(.is-divider)').element).toBeFalsy();
  });
  it('should show logout if no components are added and user is logged in', async () => {
    expect(navComponents).toHaveLength(0);
    const wrapper = createWrapper();
    setUserKind(wrapper.vm.$store, UserKinds.LEARNER);
    await wrapper.vm.$nextTick();
    expect(wrapper.findComponent(logoutSideNavEntry).element).toBeTruthy();
  });

  describe('SideNav components are shown/hidden depending on role', () => {
    afterEach(() => {
      emptyNavComponents(1);
    });
    const testCases = [
      [UserKinds.ADMIN, UserKinds.ADMIN, true],
      [UserKinds.ADMIN, UserKinds.CAN_MANAGE_CONTENT, false],
      [UserKinds.ADMIN, UserKinds.COACH, true],
      [UserKinds.ADMIN, UserKinds.LEARNER, true],
      [UserKinds.ANONYMOUS, UserKinds.ADMIN, false],
      [UserKinds.ANONYMOUS, UserKinds.ANONYMOUS, true],
      [UserKinds.ANONYMOUS, UserKinds.CAN_MANAGE_CONTENT, false],
      [UserKinds.ANONYMOUS, UserKinds.COACH, false],
      [UserKinds.ANONYMOUS, UserKinds.LEARNER, false],
      [UserKinds.CAN_MANAGE_CONTENT, UserKinds.CAN_MANAGE_CONTENT, true],
      [UserKinds.COACH, UserKinds.ADMIN, false],
      [UserKinds.COACH, UserKinds.CAN_MANAGE_CONTENT, false],
      [UserKinds.COACH, UserKinds.COACH, true],
      [UserKinds.COACH, UserKinds.LEARNER, true],
      [UserKinds.LEARNER, UserKinds.ADMIN, false],
      [UserKinds.LEARNER, UserKinds.CAN_MANAGE_CONTENT, false],
      [UserKinds.LEARNER, UserKinds.COACH, false],
      [UserKinds.LEARNER, UserKinds.LEARNER, true],
      [UserKinds.SUPERUSER, UserKinds.ADMIN, true],
      [UserKinds.SUPERUSER, UserKinds.COACH, true],
      [UserKinds.SUPERUSER, UserKinds.LEARNER, true],
      [UserKinds.SUPERUSER, UserKinds.SUPERUSER, true],
    ];

    it.each(testCases)(
      'if user is %s, then %s component should show (%s)',
      async (kind, otherKind, shouldShow) => {
        createAndRegisterComponent(`${otherKind}SideNavEntry`, url, label, icon, otherKind);
        expect(navComponents).toHaveLength(1);
        const wrapper = createWrapper();
        setUserKind(wrapper.vm.$store, kind);
        await wrapper.vm.$nextTick();
        if (shouldShow) {
          expect(wrapper.text()).toContain(label);
        } else {
          expect(wrapper.text()).not.toContain(label);
        }
      }
    );
  });

  describe('with multiple components', () => {
    afterEach(() => {
      emptyNavComponents(2);
    });
    // All user kinds that can be copresented in the side nav.
    const testCases = [
      [UserKinds.LEARNER, UserKinds.COACH],
      [UserKinds.LEARNER, UserKinds.ADMIN],
      [UserKinds.LEARNER, UserKinds.CAN_MANAGE_CONTENT],
      [UserKinds.COACH, UserKinds.ADMIN],
      [UserKinds.COACH, UserKinds.CAN_MANAGE_CONTENT],
      [UserKinds.ADMIN, UserKinds.CAN_MANAGE_CONTENT],
    ];
    it.each(testCases)('%s component should above %s component', async (kind, otherKind) => {
      createAndRegisterComponent(`${kind}SideNavEntry`, url, label, icon, kind, 10);
      createAndRegisterComponent(`${otherKind}SideNavEntry`, url, label2, icon, otherKind, 10);
      expect(navComponents).toHaveLength(2);
      const wrapper = createWrapper();
      setUserKind(wrapper.vm.$store, UserKinds.SUPERUSER);
      await wrapper.vm.$nextTick();
      const sideNavComponents = wrapper.findAll("[data-test='side-nav-component']");
      expect(sideNavComponents.exists()).toBeTruthy();
      expect(sideNavComponents.at(0).html()).toContain(label);
      expect(sideNavComponents.at(1).html()).toContain(label2);
    });
  });

  describe('and the priority flag', () => {
    afterEach(() => {
      emptyNavComponents(2);
    });

    it('should show higher priority component above lower priority component', () => {
      createAndRegisterComponent('1SideNavEntry', url, label, icon, role, 1);
      createAndRegisterComponent('2SideNavEntry', url, label2, icon, role, 10);
      expect(navComponents).toHaveLength(2);
      const wrapper = createWrapper();
      const sideNavComponents = wrapper.findAll("[data-test='side-nav-component']");
      expect(sideNavComponents.exists()).toBeTruthy();
      expect(sideNavComponents.at(0).html()).toContain(label);
      expect(sideNavComponents.at(1).html()).toContain(label2);
    });

    it('should show account section component below lower priority component', () => {
      createAndRegisterComponent('1SideNavEntry', url, label, icon, role, 1);
      createAndRegisterComponent(
        '2SideNavEntry',
        url,
        label2,
        icon,
        role,
        10,
        NavComponentSections.ACCOUNT
      );
      expect(navComponents).toHaveLength(2);
      const wrapper = createWrapper();
      const sideNavComponents = wrapper.findAll("[data-test='side-nav-component']");
      expect(sideNavComponents.exists()).toBeTruthy();
      expect(sideNavComponents.at(1).html()).toContain(label2);
      expect(sideNavComponents.at(0).html()).toContain(label);
    });

    it('should show component with priority above undefined priority component', () => {
      // Component 2 should be registered first
      createAndRegisterComponent('2SideNavEntry', url, label2, icon, role, 10);
      createAndRegisterComponent('1SideNavEntry', url, label, icon, role, undefined);
      expect(navComponents).toHaveLength(2);
      const wrapper = createWrapper();
      const sideNavComponents = wrapper.findAll("[data-test='side-nav-component']");
      expect(sideNavComponents.exists()).toBeTruthy();
      expect(sideNavComponents.at(0).html()).toContain(label2);
      expect(sideNavComponents.at(1).html()).toContain(label);
    });
  });

  describe('when on an SoUD or NOT', () => {
    describe('on an SoUD with learn-only device indicators', () => {
      let wrapper;
      beforeAll(() => {
        wrapper = createWrapper(undefined, { isLearnerOnlyImport: true });
      });
      describe('showing the SyncStatusDisplay', () => {
        it('shows the SyncStatusDisplay to Learners', async () => {
          setUserKind(wrapper.vm.$store, UserKinds.LEARNER);
          await wrapper.vm.$nextTick();
          expect(wrapper.findComponent(SyncStatusDisplay).exists()).toBe(true);
        });

        it.each([UserKinds.COACH, UserKinds.ADMIN, UserKinds.ANONYMOUS])(
          'does not show the SyncStatusDisplay to %s',
          async kind => {
            setUserKind(wrapper.vm.$store, kind);
            await wrapper.vm.$nextTick();
            expect(wrapper.findComponent(SyncStatusDisplay).exists()).toBe(false);
          }
        );
      });
      /* Note that Facilty & Coach plugins are hackily disabled in their kolibri_plugin
       * definitions - hence no tests to ensure they're hidden here when on SoUD */
      it('shows the Learn-only notice to non-Learners', async () => {
        const wrapper = createWrapper(undefined, { isLearnerOnlyImport: true });
        setUserKind(wrapper.vm.$store, UserKinds.COACH);
        await wrapper.vm.$nextTick();
        expect(wrapper.findComponent(LearnOnlyDeviceNotice).exists()).toBe(true);
        setUserKind(wrapper.vm.$store, UserKinds.ADMIN);
        await wrapper.vm.$nextTick();
        expect(wrapper.findComponent(LearnOnlyDeviceNotice).exists()).toBe(true);
      });

      it('does not show learn-only notice to Learners or Guests', async () => {
        const wrapper = createWrapper(undefined, { isLearnerOnlyImport: true });
        setUserKind(wrapper.vm.$store, UserKinds.LEARNER);
        await wrapper.vm.$nextTick();
        expect(wrapper.findComponent(LearnOnlyDeviceNotice).exists()).toBe(false);
        setUserKind(wrapper.vm.$store, UserKinds.ANONYMOUS);
        await wrapper.vm.$nextTick();
        expect(wrapper.findComponent(LearnOnlyDeviceNotice).exists()).toBe(false);
      });
    });
    describe('NOT on a SoUD', () => {
      let wrapper;
      beforeEach(async () => {
        wrapper = createWrapper(undefined, { isLearnerOnlyImport: false });
      });

      it('shows no notice', () => {
        expect(wrapper.findComponent(LearnOnlyDeviceNotice).exists()).toBe(false);
      });
    });
  });
});
