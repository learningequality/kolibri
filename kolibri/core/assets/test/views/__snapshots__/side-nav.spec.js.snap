// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`side nav component should be hidden if navShown is false 1`] = `
<div class="side-nav-wrapper" style="position: relative;">
  <div class="side-nav" style="width: 100px; color: rgb(58, 58, 58); background-color: rgb(255, 255, 255); display: none;" name="side-nav">
    <div class="side-nav-header" style="height: 20px; width: 100px; padding-top: 8px; background-color: rgb(58, 58, 58);"><button aria-label="Close navigation" type="button" tabindex="0" class="keen-ui-icon-button keen-ui-icon-button--type-secondary keen-ui-icon-button--color-white keen-ui-icon-button--size-large KeenUiIconButton-noKey-0_1e7m9ar">
        <div class="keen-ui-icon-button-icon" style="color: rgb(153, 97, 137);">
          <mat-svg name="close" category="navigation" class="side-nav-header-close"></mat-svg>
        </div>
        <div class="ui-progress-circular keen-ui-icon-button-progress ui-progress-circular--type-indeterminate" style="width: 24px; height: 24px; display: none;" name="ui-progress-circular--transition-fade"><svg role="progressbar" viewBox="25 25 50 50" aria-valuemax="100" aria-valuemin="0" class="ui-progress-circular-indeterminate" style="stroke: #03a9f4;">
            <circle cx="50" cy="50" fill="none" r="20" stroke-miterlimit="10" stroke-width="4.5" class="ui-progress-circular-indeterminate-path" style="stroke: #996189;"></circle>
          </svg></div>
        <!---->
        <!---->
      </button> <span class="side-nav-header-name" style="color: rgb(255, 255, 255);">Kolibri</span></div>
    <div class="side-nav-scrollable-area" style="top: 20px; width: 100px;">
      <div role="navigation" aria-label="Main user navigation" style="background-color: rgb(255, 255, 255);">
        <ul role="menu" class="ui-menu has-icons">
          <!---->
          <li><a role="menuitem" class="ui-menu-option is-divider">
              <!----></a></li>
          <li><a role="menuitem" class="ui-menu-option is-divider">
              <!----></a></li>
          <!---->
        </ul>
      </div>
      <div class="side-nav-scrollable-area-footer" style="color: rgb(97, 97, 97);"><img src="./kolibri-logo.svg" class="side-nav-scrollable-area-footer-logo">
        <div class="side-nav-scrollable-area-footer-info">
          <p>Kolibri testversion</p>
          <p>© 2018 Learning Equality</p>
          <p><a dir="auto" type="button" tabindex="0" class="privacy-link KButton-noKey-0_48nhro link">Usage and privacy
              <!----></a></p>
        </div>
      </div>
    </div>
  </div>
  <div class="side-nav-backdrop" style="display: none;"></div>
  <!---->
  <div dir="ltr" class="resize-sensor" style="position: absolute; left: -10px; top: -10px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;">
    <div class="resize-sensor-expand" style="position: absolute; left: -10px; top: -10px; right: 0; bottom: 0; overflow: hidden; z-index: -1; visibility: hidden;">
      <div style="position: absolute; left: 0; top: 0; transition: 0s;"></div>
    </div>
    <div class="resize-sensor-shrink" style="position: absolute; left: -10px; top: -10px; right: 0; bottom: 0; overflow: hidden; z-index: -1; visibility: hidden;">
      <div style="position: absolute; left: 0; top: 0; transition: 0s; width: 200%; height: 200%"></div>
    </div>
  </div>
</div>
`;

exports[`side nav component should show nothing if no components are added and user is not logged in 1`] = `
<div class="side-nav-wrapper" style="position: relative;">
  <div class="side-nav" style="width: 100px; color: rgb(58, 58, 58); background-color: rgb(255, 255, 255);" name="side-nav">
    <div class="side-nav-header" style="height: 20px; width: 100px; padding-top: 8px; background-color: rgb(58, 58, 58);"><button aria-label="Close navigation" type="button" tabindex="0" class="keen-ui-icon-button keen-ui-icon-button--type-secondary keen-ui-icon-button--color-white keen-ui-icon-button--size-large KeenUiIconButton-noKey-0_1e7m9ar">
        <div class="keen-ui-icon-button-icon" style="color: rgb(153, 97, 137);">
          <mat-svg name="close" category="navigation" class="side-nav-header-close"></mat-svg>
        </div>
        <div class="ui-progress-circular keen-ui-icon-button-progress ui-progress-circular--type-indeterminate" style="width: 24px; height: 24px; display: none;" name="ui-progress-circular--transition-fade"><svg role="progressbar" viewBox="25 25 50 50" aria-valuemax="100" aria-valuemin="0" class="ui-progress-circular-indeterminate" style="stroke: #03a9f4;">
            <circle cx="50" cy="50" fill="none" r="20" stroke-miterlimit="10" stroke-width="4.5" class="ui-progress-circular-indeterminate-path" style="stroke: #996189;"></circle>
          </svg></div>
        <!---->
        <!---->
      </button> <span class="side-nav-header-name" style="color: rgb(255, 255, 255);">Kolibri</span></div>
    <div class="side-nav-scrollable-area" style="top: 20px; width: 100px;">
      <div role="navigation" aria-label="Main user navigation" style="background-color: rgb(255, 255, 255);">
        <ul role="menu" class="ui-menu has-icons">
          <!---->
          <li><a role="menuitem" class="ui-menu-option is-divider">
              <!----></a></li>
          <li><a role="menuitem" class="ui-menu-option is-divider">
              <!----></a></li>
          <!---->
        </ul>
      </div>
      <div class="side-nav-scrollable-area-footer" style="color: rgb(97, 97, 97);"><img src="./kolibri-logo.svg" class="side-nav-scrollable-area-footer-logo">
        <div class="side-nav-scrollable-area-footer-info">
          <p>Kolibri testversion</p>
          <p>© 2018 Learning Equality</p>
          <p><a dir="auto" type="button" tabindex="0" class="privacy-link KButton-noKey-0_48nhro link">Usage and privacy
              <!----></a></p>
        </div>
      </div>
    </div>
  </div>
  <div class="side-nav-backdrop"></div>
  <!---->
  <div dir="ltr" class="resize-sensor" style="position: absolute; left: -10px; top: -10px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;">
    <div class="resize-sensor-expand" style="position: absolute; left: -10px; top: -10px; right: 0; bottom: 0; overflow: hidden; z-index: -1; visibility: hidden;">
      <div style="position: absolute; left: 0; top: 0; transition: 0s;"></div>
    </div>
    <div class="resize-sensor-shrink" style="position: absolute; left: -10px; top: -10px; right: 0; bottom: 0; overflow: hidden; z-index: -1; visibility: hidden;">
      <div style="position: absolute; left: 0; top: 0; transition: 0s; width: 200%; height: 200%"></div>
    </div>
  </div>
</div>
`;
